### PR TITLE
TST Ignore floating point noise in gradient checkpointing test

### PR DESCRIPTION
**Pull request by jiqing-feng**
## Why

`_test_training_gradient_checkpointing` compares which parameters received a gradient with
and without gradient checkpointing, using an exact-zero check (`p.grad.abs().sum() > 0`).

Some parameters only receive gradients that are mathematically zero. Floating point errors
can turn those into tiny non-zero values, and whether they do depends on the reduction order
inside the kernels, which is not deterministic on all devices. The check then classifies the
same parameter differently in the two passes.

This makes the test flaky on XPU: 8 to 10 of 444 parametrizations fail per run, a different
set every time, across many tuner methods, both `use_reentrant` values and several models.

Gradient checkpointing is not involved. Running the same forward and backward twice *without*
checkpointing already differs on XPU:

| device | two identical passes agree | value of a noise gradient |
| --- | --- | --- |
| CPU | 20/20 runs | stable `1.8e-05`, never zero |
| XPU | 18/20 runs | `2e-06` to `1e-05`, occasionally exactly `0.0` |

## The fix

Run the reference pass twice and ignore gradients that are either negligible compared to the
largest gradient or not reproducible. Both criteria are needed:

- Gemma3 + LoRA: noise is `~5e-06` against a real gradient of `~4e+03`, but the two passes
  sometimes agree, so only the threshold catches it.
- OPT + VBLoRA: `vblora_logits_*` varies by more than 3x between passes (`3.7e-05` vs
  `1.1e-05`) while not being small enough relatively, so only the reproducibility check
  catches it.

On CPU and CUDA gradients are reproducible and the noise gradients are exactly `0.0`, so
nothing is ignored and the behaviour is unchanged. A parameter that is consistently zero
without checkpointing and non-zero with it is still caught.

## Validation

8x Intel Arc Pro B60, `torch 2.13.0+xpu`:

- `test_training_decoders_gradient_checkpointing`, 6 consecutive runs: 394 passed, 50 skipped
  every time (8 to 10 failures per run before)
- encoder-decoder and feature-extraction suites, `-k gradient_checkpointing`: 306 passed,
  82 skipped

### CODE REVIEW

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for explaining the issue, it is a bit strange to see non-zero gradients where they should be zero, but it is how it is.

About the fix, I was wondering if it needs to be so  complicated. When we have:

`non_zero_grad_params_normal = {n for n, p in params if p.grad.abs().sum() > 0}`

Would it not be enough to increase the threshold from 0 to some tiny number? If this leads to false negatives, we could try increasing the loss signal. As I don't have access to an XPU, I can't test that suggestion. With CUDA, I had no issue in 1480 test runs.

**Comment by jiqing-feng** [comment]
Thanks for the review. I tried your suggestion, but the data says a single threshold does not work. Here is what I measured on XPU.

## Why a threshold does not work

I dumped the full gradient dict of both invocations for every parametrization over 12 runs of `test_training_decoders_gradient_checkpointing` (4728 invocation pairs) and swept the threshold, relative to the largest gradient of the same test case:

| threshold | mismatching pairs (of 4728) |
| --------- | --------------------------- |
| 0 (current) | 94 |
| 1e-8 | 19 |
| 3e-8 | 2 |
| 1e-7 | 3 |
| 1e-6 | 11 |
| 1e-5 | 61 |

No value reaches 0. Too low and noise gets through, too high and real gradients get cut (`GloraConfig` on OPT loses 61 pairs at 1e-5). An absolute constant is worse still, gradients across the suite span 4e-14 to 5e+06.

The reason is that the two populations overlap. In `VBLoRAConfig` + `tiny-random-OPTForCausalLM` the ratios form a continuum down to 1e-10, so any cut lands inside a populated region.

I also tried, and rejected:

- comparing gradient values with `assert_close`: 732/4728 failures even at `rtol=0.5`, the two code paths simply produce different values on these tiny random models
- `p.grad is not None`: never true for any parameter in any of the 394 cases, so the assert would be vacuous
- `torch.use_deterministic_algorithms(True)` via the existing `@require_deterministic_for_xpu`: 4 failures in 16 runs, spread over 4 configs. It removes run-to-run jitter but not the difference between the two code paths
- weighting the loss to break the cancellation: 6/15/13 failures, worse than baseline

## What the current version does

Rather than requiring the two sets to be equal, only require that a gradient which is clearly non-zero in one invocation is non-zero in the other one too. The bound for "clearly non-zero" and the bound for "passes" are deliberately different, so the range where noise and real gradients overlap is not asserted on at all.

0 mismatches on the same 4728 pairs, and it does not depend on the constant: every value from 1e-6 to 1e-3 gives 0. Locally 8/8 consecutive full runs are green (394 passed, 50 skipped), plus the encoder-decoder and feature-extraction gradient checkpointing tests (306 passed, 82 skipped).

**Comment by jiqing-feng** [comment]
Thanks for the review. I tried raising the threshold first, but I don't think a single threshold can work here, for a reason that also explains why the current check is asymmetric.

## Why a single threshold can't work

A single threshold assumes noise and real gradients are two separated populations, so that some cut lies between them. They aren't. The residue of a mathematically-zero gradient is proportional to the magnitude of the terms that cancelled, which varies per parameter, so the two populations overlap.

I swept the threshold over 4728 invocation pairs collected from 12 full runs of the test on XPU. Best case is 2 mismatches at `3e-8`, and it degrades in both directions (19 at `1e-8`, 11 at `1e-6`, 61 at `1e-5`). No value reaches 0, and the good region is too narrow to rely on anyway.

## Why the asymmetric check works

The test checks whether a gradient *reaches* a parameter, not whether its value is right. That distinction is what makes the two sides asymmetric:

- If the checkpointing path is genuinely broken, backward never reaches the parameter, so `grad` is `None` or exactly `0.0`. Nothing was computed, so there is no residue.
- A tiny value like `1e-9` is itself evidence that backward *did* reach the parameter. It is the leftover of real terms cancelling.

So `> 0` is not a loosened version of the check, it is the correct predicate for what the test asserts. The threshold is only used to decide *whether to check at all*, which keeps the band where noise and real gradients overlap out of the assertion entirely. That is also why applying the threshold to both sides doesn't work: it collapses back into the single-threshold case above.

Comparing values with `assert_close` isn't an option either, the two paths produce genuinely different gradients on these tiny random models (732/4728 pairs differ even at `rtol=0.5`).

Result: 0 mismatches on the same 4728 pairs, and unlike the single threshold it doesn't depend on the constant, anything from `1e-6` to `1e-3` gives 0. 8/8 consecutive full runs green locally.

**Comment by BenjaminBossan** [comment]
Okay, so a single threshold does not work, thanks for explaining. Still I'm unsure about the solution:

```python
            # zero gradients can show up as floating point noise, so only check the clearly non-zero ones
            threshold = max([*grads_normal.values(), *grads_checkpointing.values()], default=0.0) * 1e-6
            for n in grads_normal:
                if grads_normal[n] > threshold:
                    assert grads_checkpointing[n] > 0, n
                if grads_checkpointing[n] > threshold:
                    assert grads_normal[n] > 0, n
```

So we create a threshold that, for this specific case, should delineate noise from signal, right? But why do we then assert that the gradient at the detected point is > 0? Does this not mean that we could accidentally assert True even if it's just noise? I would have thought that the threshold needs to be applied in both directions.

Maybe I'm still misunderstanding, but then the comment needs to be extended.

Btw. if the test flaky and generally passes on retry, we could also just add that (say, 3 retries, with a comment to explain why). It would be less efficient, but at least we don't need to make the test too complicated.

**Comment by jiqing-feng** [comment]
Extended the comment in 33779d0d.

**The threshold does not separate noise from signal.** It only decides whether a parameter is checked at all. That is a weaker job, which is why a single constant can do it while no single constant can do the classification.

**`> 0` is the right check because of what the test asserts**: that the gradient reaches the parameter, not that its value is correct. If checkpointing were broken, backward would not reach the parameter and the gradient would be exactly 0, because nothing was computed. A tiny value like `1e-9` is the rounding residue of terms that cancelled, so it is evidence that backward *did* reach the parameter. Noise asserting True is therefore not a false pass, the noise is itself the signal being tested for.

**Applying the threshold on both sides is the single-threshold approach**, just written differently. It puts the ambiguous range back into the assertion, and as measured earlier no value for it works.

The test cannot check gradient values instead: the two paths produce genuinely different gradients on these tiny random models (732/4728 pairs differ even at `rtol=0.5`).

On retries: they would work, but they hide the flakiness rather than remove it, and the failing parametrizations are not uniformly distributed, some fail often enough that 3 retries still leaves a visible failure rate. Happy to switch to that if you prefer it, though.

File: tests/testing_common.py [hunk 1]
```diff
@@ -1248,29 +1248,38 @@ def _test_training_gradient_checkpointing(self, model_id, config_cls, config_kwa
...
-
-            # invocation with gradient checkpointing for comparison
+            def get_grads():
+                for _, param in params:
+                    param.grad = None
+                # note we're squaring the output for bigger gradients
+                output = model(**inputs)[0] ** 2
+                loss = output.sum()
+                loss.backward()
+                return {n: param.grad.abs().sum().item() for n, param in params}
```
Comment (BenjaminBossan) [inline tests/testing_common.py]: May need a check that `.grad` is not `None`.

File: tests/testing_common.py [hunk 2]
```diff
@@ -1248,29 +1248,38 @@ def _test_training_gradient_checkpointing(self, model_id, config_cls, config_kwa
...
+
+            # What is being checked is that the gradient reaches the parameter at all, not that its value is correct.
+            # If gradient checkpointing were broken, backward would not reach the parameter and the gradient would be
+            # exactly 0. A tiny value is instead the rounding residue of terms that cancelled, i.e. evidence that
+            # backward did reach the parameter, hence `> 0` is the right check. Whether such a residue ends up as 0 or
+            # as, say, 1e-9 depends on the reduction order inside the kernels and is not deterministic on all devices.
+            # The threshold therefore only decides whether a parameter is checked at all, keeping the range where noise
+            # and real gradients overlap out of the assertion. Using the threshold on both sides would put that
+            # ambiguous range back in, and no single value for it separates noise from signal for all parameters.
+            threshold = max([*grads_normal.values(), *grads_checkpointing.values()], default=0.0) * 1e-6
```
Comment (BenjaminBossan) [inline tests/testing_common.py]: Just wondering if mean or median (with a smaller scale) would be better, as they're more robust than max.

File: tests/testing_common.py [hunk 3]
```diff
@@ -1248,29 +1248,38 @@ def _test_training_gradient_checkpointing(self, model_id, config_cls, config_kwa
...
+            grads_checkpointing = get_grads()
+
+            # What is being checked is that the gradient reaches the parameter at all, not that its value is correct.
+            # If gradient checkpointing were broken, backward would not reach the parameter and the gradient would be
+            # exactly 0. A tiny value is instead the rounding residue of terms that cancelled, i.e. evidence that
+            # backward did reach the parameter, hence `> 0` is the right check. Whether such a residue ends up as 0 or
+            # as, say, 1e-9 depends on the reduction order inside the kernels and is not deterministic on all devices.
+            # The threshold therefore only decides whether a parameter is checked at all, keeping the range where noise
+            # and real gradients overlap out of the assertion. Using the threshold on both sides would put that
+            # ambiguous range back in, and no single value for it separates noise from signal for all parameters.
```
Comment (BenjaminBossan) [inline tests/testing_common.py]: How about something along the lines of:

> A gradient of 0 does not prove the gradient was never computed: a sum of terms that cancel may come out as exactly 0 or as a tiny residue, depending on non-deterministic properties of the accelerator (observed on XPU). Therefore, only check parameters whose gradient is substantially different from 0 on at least one side, and on the other side only require that it is non-zero, since any non-zero value proves that backward reached the parameter.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for providing further information. I added some comments to the PR, but the general direction is good.

There is also a merge conflict, as a new PEFT method, ShadowPEFT, can end up with `grad = None`. When resolving the merge conflict, ensure that this case is taken care of. I think it only affects one line in the new code that I commented on, but please double check that ShadowPEFT passes the test.

**Comment by jiqing-feng** [comment]
Merged main and addressed all four points.

**`grad is None`**: `get_grads` now maps `None` to `0.0`. This is actually exercised, ShadowPEFT on Gemma3 has 8 of 34 trainable parameters with `grad is None`. Both sides are 0 for those, so they never enter an assertion, which matches the intent of the `ShadowConfig` skip further down.

**mean instead of max**: done. Measured both over the 4728 invocation pairs, they are equivalent here (0 mismatches at every scale from `1e-6` to `1.0`), so switching costs nothing.

Median does not work, though: 59 mismatches over 21 test cases, at every scale. In most cases the majority of parameters carry only noise gradients, so the median already sits in the noise range (often exactly 0), which collapses the threshold to ~0 and restores the original flakiness. Lowering the scale cannot help, which is why the number is constant across scales.

**Comment**: used your wording.

**Validation** (8x Intel Arc Pro B60, `torch 2.13.0+xpu`), all after the merge:

- `test_decoder_models.py -k gradient_checkpointing`, 5 consecutive runs: 426 passed, 79 skipped every time
- the 12 `ShadowConfig` parametrizations of that test: 12 passed
- `test_encoder_decoder_models.py` and `test_feature_extraction_models.py -k gradient_checkpointing`: 318 passed, 82 skipped (neither collects `ShadowConfig`)
- `test_shadow.py`: 40 passed


@BenjaminBossan

**Comment by BenjaminBossan** [comment]
@jiqing-feng Thanks for the further test. Could you please merge with the latest main branch, this should resolve the code quality issue.

**Comment by jiqing-feng** [comment]
> @jiqing-feng Thanks for the further test. Could you please merge with the latest main branch, this should resolve the code quality issue.

Merged.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the update, the PR now LGTM.
